j5_zoloto/component.py

```diff
@@ -0,0 +1,82 @@
+"""Classes for Fiducial Marker Camera."""
+
+from abc import abstractmethod
+from pathlib import Path
+from typing import Generator, List, Type, Union
+
+from j5.components.component import Component, Interface
+from zoloto.marker import BaseMarker, EagerMarker, Marker, UncalibratedMarker
+
+
+class MarkerCameraInterface(Interface):
+    """An interface containing the methods required for a marker camera."""
+
+    def process_frame(
+        self,
+        identifier: int,
+    ) -> Generator[Union[UncalibratedMarker, Marker], None, None]:
+        """
+        Get markers that the camera can see.
+
+        :param identifier: Camera identifier, ignored.
+        """
+        raise NotImplementedError  # pragma: nocover
+
+    def process_frame_eager(self, identifier: int) -> Generator[EagerMarker, None, None]:
+        """
+        Get markers that the camera can see.
+
+        :param identifier: Camera identifier, ignored.
+        """
+        raise NotImplementedError  # pragma: nocover
+
+    @abstractmethod
+    def save_annotated_image(self, identifier: int, file: Path) -> None:
+        """Save an annotated image to a file."""
+        raise NotImplementedError  # pragma: nocover
+
+
+class MarkerCamera(Component):
+    """
+    Camera that can identify fiducial markers.
+
+    Additionally, it will do pose estimation, along with some calibration
+    in order to determine the spatial positon and orientation of the markers
+    that it has detected.
+    """
+
+    def __init__(
+            self,
+            identifier: int,
+            backend: MarkerCameraInterface,
+    ) -> None:
+        self._backend = backend
+        self._identifier = identifier
+
+    @staticmethod
+    def interface_class() -> Type[MarkerCameraInterface]:
+        """Get the interface class that is required to use this component."""
+        return MarkerCameraInterface
+
+    @property
+    def identifier(self) -> int:
+        """An integer to identify the component on a board."""
+        return self._identifier
+
```

Comment:
Thoughts on exposing the `get_visible_markers` method? It's faster than the main marker, and for many cases is much easier to use.

Reply:
d72bb3c

Reply:
Not completely sure about the method name to expose to students here.